### ISSUE TITLE
Question about fine-tuning

### ISSUE THREAD

**Issue by tulvgengenr** [mahmoodlab/TITAN#22]
Very great work! I have a question for you, assuming that I now have my own private WSIs and corresponding pathology reports, how should I fine-tune or post-train TITAN so as to achieve better performance in a specific area?

**Comment by sophiajw** [comment]
Hi @tulvgengenr, 

thanks for your question! I just added a script for fine-tuning TITAN on a supervised downstream task (`titan/finetune.py`). It includes the hyperparameter choices that we have used for supervised fine-tuning experiments in our paper as default values. You just need to adapt the data class to your custom task to be able to run it on a downstream task. We have not investigated fine-tuning on paired WSIs and reports but you could achieve this in a similar way. 

Hope this helps!
Best
Sophia

**Comment by tulvgengenr** [comment]
Thank you very much for your reply, I would like to ask if the code for stage3 pre-training will be released, i.e. to do the alignment between the WSI level feature and the pathology report?

**Comment by andrewsong90** [comment]
Hi @tulvgengenr 

Stage 3 is CoCa, which you can adapt it from the original codebase.
But yes, upon publication, we will include it.

Best,
Andrew

**Comment by tulvgengenr** [comment]
> Hi [@tulvgengenr](https://github.com/tulvgengenr)
> 
> Stage 3 is CoCa, which you can adapt it from the original codebase. But yes, upon publication, we will include it.
> 
> Best, Andrew

Thanks for your reply @andrewsong90 
I have now reproduced CoCa's algorithm, but I have a question while training, for different WSIs, the number of their patches is different, how should I go about the padding operation in order to do batch training?

**Comment by andrewsong90** [comment]
Hi @tulvgengenr 

For Stage 3 pretraining (CoCa with WSIs and reports), we operate on a larger feature grid of 64 x 64 (explained further in online methods of our preprint). We observe this usually does the job of covering WSIs. For padding, we zero-pad and then use attention-masking (for training), such that 64x64 feature grid is always presented.

Hope this helps,
Andrew

**Comment by tulvgengenr** [comment]
> Hi [@tulvgengenr](https://github.com/tulvgengenr)
> 
> For Stage 3 pretraining (CoCa with WSIs and reports), we operate on a larger feature grid of 64 x 64 (explained further in online methods of our preprint). We observe this usually does the job of covering WSIs. For padding, we zero-pad and then use attention-masking (for training), such that 64x64 feature grid is always presented.
> 
> Hope this helps, Andrew

Hi @andrewsong90 

I am now fine-tuning based on my own private pathology dataset (which contains pathology reports), I am currently performing the following steps: 
1. Segmentation of the background is performed through CLAM, which slices the WSI into patches. 
2. Feature extraction is performed on the patches through Conch v1.5, which yields patch-level features.
3. Following the prompt used by TITAN, the pathology report was preprocessed using Qwen2-7B-Instruct.
4. The code for CoCa contrast learning was reproduced and trained using the contrast learning loss, following https://github.com/lucidrains/CoCa-pytorch

The main code of step 4 is as follows, which I added as a forward function to the TITAN Module:
```python
def forward(
        self,
        patch_features,
        patch_coords,
        patch_size_lv0,
        texts,
    ):
        '''
        Adapted from https://github.com/lucidrains/CoCa-pytorch/blob/main/coca_pytorch/coca_pytorch.py
        '''
        batch, device = len(patch_features), patch_features[0].device
        
        input_ids = self.text_encoder.tokenizer(texts).to(device)

        text_latents = self.encode_text(input_ids, normalize=True).to(device)

        wsi_latents = self.encode_slide_from_patch_features(
            patch_features, patch_coords, patch_size_lv0
        )

        # calculate contrastive loss
        ce = F.cross_entropy
        sim = einsum('i d, j d -> i j', text_latents, wsi_latents)
        contrastive_labels = torch.arange(batch, device=device)

        contrastive_loss = (ce(sim, contrastive_labels) + ce(sim.t(), contrastive_labels)) * 0.5
        contrastive_loss = contrastive_loss

        return contrastive_loss
```

But my model is not converging at all, even I found that the checkpoint initialisation weights of the pre-trained model using TITAN are worse than the direct initialisation of the model in loss value.
May I ask if you use such CoCa training method when training in Stage3? Or where do you think there might be a problem with my process?
